mavros_extras/src/plugins/fake_gps.cpp

```diff
@@ -0,0 +1,309 @@
+/**
+ * @brief Fake GPS with local position source plugin
+ * @file fake_gps.cpp
+ * @author Christoph Tobler <toblech@ethz.ch>
+ * @author Nuno Marques <n.marques21@hotmail.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2015 Christoph Tobler.
+ * Copyright 2017 Nuno Marques.
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <mavros/mavros_plugin.h>
+#include <mavros/setpoint_mixin.h>
+#include <eigen_conversions/eigen_msg.h>
+#include <GeographicLib/Geocentric.hpp>
+#include <GeographicLib/Geoid.hpp>
+
+#include <geometry_msgs/PoseStamped.h>
+#include <geometry_msgs/TransformStamped.h>
+
+namespace mavros {
+namespace extra_plugins {
+using mavlink::common::GPS_FIX_TYPE;
+
+/**
+ * @brief Fake GPS plugin.
+ *
+ * Sends fake GPS from local position estimation source data (motion capture,
+ * vision) to FCU - processed in HIL mode or out of it if parameter MAV_USEHILGPS
+ * is set on PX4 Pro Autopilot Firmware; Ardupilot Firmware already supports it
+ * without a flag set.
+ */
+class FakeGPSPlugin : public plugin::PluginBase,
+	private plugin::TF2ListenerMixin<FakeGPSPlugin> {
+public:
+	FakeGPSPlugin() : PluginBase(),
+		fp_nh("~fake_gps"),
+		gps_period(0.2),
+		use_mocap(true),
+		map_origin(0.0, 0.0, 0.0),
+		mocap_transform(true),
+		use_vision(false),
+		tf_listen(false),
+		tf_rate(10.0),
+		eph(2.0),
+		epv(2.0),
+		satellites_visible(5),
+		fix_type(GPS_FIX_TYPE::NO_GPS)
+	{ }
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		double _gps_period;
+		double origin_lat, origin_lon, origin_alt;
+
+		last_pos_time = ros::Time(0.0);
+
+		// general params
+		int ft_i;
+		fp_nh.param<int>("fix_type", ft_i, utils::enum_value(GPS_FIX_TYPE::NO_GPS));
+		fix_type = static_cast<GPS_FIX_TYPE>(ft_i);
+		fp_nh.param("gps_period", _gps_period, 0.2);		// GPS data rate of 5hz
```

Comment:
Well bellow it is used to add throttle to the msg (I kept Thomas version on that). But I suppose that can br changed.